### DISCUSSION TITLE
[@types/react] Missing dependency declaration

### DISCUSSION THREAD

**Discussion by DerGary** [DefinitelyTyped/DefinitelyTyped#57363] [Issues with a @types package]
When using yarn2 with PNP and webpack, module resolution is a lot stricter. Currently I get the following Warnings because of missing dependency declaration:

```
(node:35) [MODULE_NOT_FOUND] Error: @types/react tried to access prop-types, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react tried to access scheduler, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-transition-group tried to access react, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-redux tried to access react, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/hoist-non-react-statics tried to access react, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-router-dom tried to access react-router, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-router-dom tried to access react, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-router-dom tried to access history, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-router tried to access react, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:35) [MODULE_NOT_FOUND] Error: @types/react-router tried to access history, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
(node:42) [MODULE_NOT_FOUND] Error: @types/react-dom tried to access react, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
```

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Thanks for the report!

Practically `@types/react` is doing the correct thing and listing `@types/prop-types` and `@types/scheduler` as direct dependencies. Yarn just doesn't recognize that they're type dependencies. In a normal project you wouldn't import from `@types/react` either so I feel like there's  a missing piece on Yarn's side. 
@arcanis How should `@types` packages declare type dependencies?

**Reply by arcanis** [reply]
I suspect the OP is using the optional [loose mode](https://yarnpkg.com/features/pnp#pnp-loose-mode), which emits warnings on "invalid accesses" rather than outright crashing. The problem is that some packages (and TypeScript is one of them) use `try/catch` to silence resolution errors and, since the loose mode uses warning rather than exceptions, those cannot be catched (we document this unfortunate behaviour in more details [here](https://yarnpkg.com/features/pnp#caveat)). It's not a problem in terms of functional behaviour, but it causes a bunch of spurious warnings.

In any case, `@types/react` is properly defined, so nothing to do on your side. Perhaps Yarn can implements some kind of mitigation to tolerate `@types` module resolution errors.

**Reply by DerGary** [reply]
Hello @eps1lon, 

I don't see where @types/react lists any dependencies apart from csstypes in its package.json: 
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/package.json :
```
{
  "private": true,
  "dependencies": {
    "csstype": "^3.0.2"
  }
}
```

further when I'm right its not about the `@types/prop-types` but `prop-types` itself which is used by `@types/react` for example here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts

```
import * as CSS from 'csstype';
import * as PropTypes from 'prop-types';
import { Interaction as SchedulerInteraction } from 'scheduler/tracing';
```

@arcanis It's right I'm using loose mode. I migrated these days from yarn with node_modules resolver to yarn PNP and without loose mode I wasn't able to run my application. But I will try without it!.

**Reply by eps1lon** [reply]
> [`master`/types/react/package.json](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/package.json) :



> I don't see where @types/react lists any dependencies apart from csstypes in its package.json:
> [`master`/types/react/package.json](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/package.json) :

The source code is not relevant here since the types-publisher adds the `@types/` dependencies: https://unpkg.com/@types/react@17.0.37/package.json

> further when I'm right its not about the @types/prop-types but prop-types itself which is used by @types/react for example here:

Yes, this is how `@types/` packages are used in any code. You never write `import { SomeType } from '@types/react'` but `import { SomeType } from 'react'`.

As @arcanis (the lead maintainer of Yarn Berry) already said: `@types/react` is doing everything right. It's TypeScript that needs adjustment.

**Reply by DerGary** [reply]
thanks for the clarification, I didn't knew that.

fyi I tried strict mode but got errors from at least 3 different packages none of which are @types packages. So I can't use strict mode until these errors get resolved by the maintainers of these packages.

**Reply by arcanis** [reply]
This is starting to go outside of the scope of the DefinitelyTyped repo, but you can fix them on your side using [`packageExtensions`](https://yarnpkg.com/configuration/yarnrc#packageExtensions). Feel free to ask on the Yarn Discord if you have further questions on how to use the strict mode 🙂
